stdlib/LinearAlgebra/src/matmul.jl

```diff
@@ -1007,3 +1007,50 @@ function matmul3x3!(C::AbstractMatrix, tA, tB, A::AbstractMatrix, B::AbstractMat
     end # inbounds
     C
 end
+
+# Three-argument *
+*(x::AdjointAbsVec{<:Number}, A::AbstractMatrix, y::AbstractVector) = dot(x.parent,A,y)
+*(x::TransposeAbsVec{<:Real}, A::AbstractMatrix, y::AbstractVector) = dot(x.parent,A,y)
```

Comment:
I guess we'll need a "3-arg `_dot_nonrecursive`" mimicking the generic 3-arg `dot`, but calling `*` in the loop instead of `dot` recursively. Perhaps roughly as
```julia
function _3argmul(f, x, A, y, transform=identity)
    ...
end

dot(x::AbstractVector, A::AbstractMatrix, y::AbstractVector) = _3argmul(dot, x, A, y)
*(x::Adjoint{<:Any,<:AbstractVector}, A::AbstractMatrix, y::AbstractVector) = _3argmul(*, x.parent, A, y, adjoint)
*(x::Transpose{<:Any,<:AbstractVector}, A::AbstractMatrix, y::AbstractVector) = _3argmul(*, x.parent, A, y)
```
inspired by the approach in
https://github.com/JuliaLang/julia/blob/4ccfd37dc9c593a1b208537fbd0a26b6dfaaf37d/stdlib/LinearAlgebra/src/diagonal.jl#L708-L714
This could be further specialized for tridiagonal matrices etc, as is `dot`.

Reply:
I'd really like to avoid having to specialise on all these special types. See what you think of `_mat_mat_vec`, which should only get called if you miss all the more specific methods.